### PR TITLE
Add deploy container with compiled executables

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

This should allow running CCE on clusters with Singularity without having to compile anything. I haven't tested Singularity recently though.

Try it like this (once merged):

```
docker run sxs-collaboration/spectre CharacteristicExtract --input-file path/to/input/file
```

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: containers/Dockerfile.buildenv [hunk 1]
```diff
@@ -354,21 +354,50 @@ RUN apt-get -y clean
...
+    -D CMAKE_Fortran_COMPILER=gfortran \
+    -D CMAKE_BUILD_TYPE=Release \
+    -D DEBUG_SYMBOLS=OFF \
+    -D BUILD_PYTHON_BINDINGS=ON \
+    -D MEMORY_ALLOCATOR=SYSTEM \
+    ..
+
+# Skip compiling the executables if we are emulating the target architecture
+# because that takes a long time. They can be compiled on-demand.
+RUN if [ "$TARGETARCH" = "$BUILDARCH" ] ; then \
```
Comment (knelli2) [inline containers/Dockerfile.buildenv:386]: Should we just only offer deploy on x86? I don't know of any arm clusters we'll be running on.

File: containers/Dockerfile.buildenv [hunk 2]
```diff
@@ -354,21 +354,50 @@ RUN apt-get -y clean
...
+# Deploy compiled executables to an image that can be run on HPC systems.
+# - We could also compile in the dev container and then copy the executables to
+#   a minimal deploy container to reduce its size. That requires keeping track
+#   of the shared libraries that are needed to run the executables, so to make
+#   things easier we just base the deploy container on the dev container.
+# - We inherit from the remote image rather than the local dev because it is
+#   faster on release CI to just pull the remote dev image rather than having to
+#   build the dev container again. We could pre-fetch the dev image and then
+#   build from the local cache, but that didn't work right away.
+FROM sxscollaboration/spectre:dev AS deploy
```
Comment (knelli2) [inline containers/Dockerfile.buildenv:366]: Are we going to have different `deploy` containers for different executables? Or just one all executables we'll offer? I could also see this container being called `cce` instead so it's more clear that's what it is.

File: containers/Dockerfile.buildenv (hunk 1 above)
Comment (nilsvu) [reply containers/Dockerfile.buildenv:386]: The demo container is based on this, and that's built for ARM.

File: containers/Dockerfile.buildenv (hunk 2 above)
Comment (nilsvu) [reply containers/Dockerfile.buildenv:366]: I thought we go with one deploy container for now to keep things simple. If it turns out the container increases in size too much we can split it up.

**Comment by nilsvu** [comment]
@knelli2 could you take another look at this please?

**Review by knelli2 — APPROVED** [review]

**Comment by knelli2** [comment]
Ignoring known CI issues

**Pull request merged**